### PR TITLE
feat: (Core) migrate popper.js to cdk

### PR DESCRIPTION
**Pull request by JKMarkowski**
#### Please provide a brief summary of this pull request.

There is introducement of new popover, based on CDK Overlay, which will become default one. 
New one has been introduced for every component, which were using old popover.

API Changes:
Deprecated:
- options (old popper.js option reference - removed)
- addContainerClass (moved to additionalBodyClass)
- additionalClasses (moved to additionalTriggerClass)

From new features:
- Different ways of handling scroll effects - @Input() scrollStrategy
- New way to add CDK's placement - @Input() cdkPositions
- @Input() focusAutoCapture - To automatically move focus into popover region.
- Internal RTL support - no need to change placement by user.
- @Input() fixedPosition - to keep same position (Otherwise some default callback positions will be added, to prevent from overflowing the window)

Some components have migrated focusTrap to CDK's focusTrap

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [ ] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

File: libs/core/src/lib/form/form-input-message-group/form-input-message-group.component.html [hunk 1]
```diff
@@ -8,8 +8,7 @@
     [closeOnOutsideClick]="closeOnOutsideClick"
     [isOpen]="isOpen"
     (isOpenChange)="openChanged($event)"
-    [addContainerClass]="'fd-popover-container-custom--message'"
-    [additionalClasses]="['fd-popover__popper--input-message-group']"
+    [additionalBodyClass]="'fd-popover__popper--input-message-group'"
```
Comment (Betrozov) [inline libs/core/src/lib/form/form-input-message-group/form-input-message-group.component.html:11]: unnecessary binding

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+    ViewportRuler
+} from '@angular/cdk/overlay';
+import { RtlService } from '../../utils/services/rtl.service';
+import { BasePopoverClass } from '../base/base-popover.class';
+import { TemplatePortal } from '@angular/cdk/portal';
+import { merge, Observable, Subject } from 'rxjs';
+import { distinctUntilChanged, filter, startWith, takeUntil } from 'rxjs/operators';
+import { ConnectedOverlayPositionChange, ConnectionPositionPair } from '@angular/cdk/overlay/position/connected-position';
+import { DefaultPositions, PopoverPosition } from './popover-position';
+import { KeyUtil } from '../../utils/functions/key-util';
```
Comment (Betrozov) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:35]: Sort imports (angular → 3rd party or aliases → relative)

File: libs/core/src/lib/popover/popover.module.ts [hunk 3]
```diff
@@ -11,6 +11,9 @@ import { PopoverBodyHeaderDirective } from './popover-body/popover-body-directiv
 import { PopoverBodyFooterDirective } from './popover-body/popover-body-directives/popover-body-footer.directive';
 
 import { ButtonModule } from '../button/button.module';
+import { OverlayModule } from '@angular/cdk/overlay';
+import { CdkPopoverComponent } from './cdk-overlay/cdk-popover.component';
+import { A11yModule } from '@angular/cdk/a11y';
```
Comment (Betrozov) [inline libs/core/src/lib/popover/popover.module.ts:16]: plz sort imports (angular → 3rd party or aliases → relative)

**Review by Betrozov — CHANGES_REQUESTED** [review]

**Comment by Betrozov** [comment]
plz fix travis

**Comment by mikerodonnell89** [comment]
Combobox - hovered item does not take up whole popover
Screen Shot 2020-10-12 at 3 57 23 PM
 width

**Comment by mikerodonnell89** [comment]
Multi-input gets attached to the top left of the page

Screen Shot 2020-10-12 at 3 58 34 PM

**Comment by mikerodonnell89** [comment]
Datepicker and datetimepicker also get attached to the top left, and the latter throws a console error

Screen Shot 2020-10-12 at 3 59 26 PM

**Comment by mikerodonnell89** [comment]
Form message a few px off the input (and the input is shorter?)

Screen Shot 2020-10-12 at 4 08 18 PM

**Comment by mikerodonnell89** [comment]
Message strip

Screen Shot 2020-10-12 at 4 10 03 PM

**Comment by mikerodonnell89** [comment]
"fixed position" popover example not staying fixed

**Comment by mikerodonnell89** [comment]
Shellbar popover right-aligned

Screen Shot 2020-10-12 at 4 14 23 PM

**Comment by mikerodonnell89** [comment]
TimePicker throws console error when popover opened

**Comment by JKMarkowski** [comment]
> Shellbar popover right-aligned
> 
> Screen Shot 2020-10-12 at 4 14 23 PM

It's bottom end, so it's correct

**Comment by JKMarkowski** [comment]
> "fixed position" popover example not staying fixed

Could you write something more about it?

**Comment by JKMarkowski** [comment]
> Message strip
> 
> Screen Shot 2020-10-12 at 4 10 03 PM

@mikerodonnell89 Can't reproduce it anymore

**Comment by JKMarkowski** [comment]
> Combobox - hovered item does not take up whole popover
> Screen Shot 2020-10-12 at 3 57 23 PM
> width



> Form message a few px off the input (and the input is shorter?)
> 
> Screen Shot 2020-10-12 at 4 08 18 PM

@mikerodonnell89 We need to discuss about it on meeting

**Comment by InnaAtanasova** [comment]
Don't know if it's related to this PR but in Combobox, clicking anywhere on the page focuses the first input field of the first example and the menus of the other examples cannot be open.

**Comment by InnaAtanasova** [comment]
Screen Shot 2020-10-19 at 8 54 32 AM

**Comment by InnaAtanasova** [comment]
Not sure if it was intentional:
Screen Shot 2020-10-19 at 8 52 54 AM
Screen Shot 2020-10-19 at 8 52 58 AM



This is broken:
Screen Shot 2020-10-19 at 8 53 09 AM

**Comment by katekozlowska** [comment]
Popover in dialog example - expanded list is covered by overlay, see screenshot

deploy-preview-3582--fundamental-ngx netlify app_

**Comment by katekozlowska** [comment]
Incorrect styling in Popover CDK Placement (with particular set on attributes)

deploy-preview-3582--fundamental-ngx netlify app_ (1)

File: apps/docs/src/app/core/component-docs/popover/popover-docs.component.html [hunk 4]
```diff
@@ -99,25 +186,13 @@
...
-
-<description>
-    <p>
-        Under the hood, the Popover component makes use of <code>PopoverDirective</code>. It abstracts some
-        functionality in favor of consistency, but it may not be sufficient for every use case. If more flexibility is
-        required, we recommend the use of the directive which can be found
-        <a [routerLink]="['/popover-directive']">here</a>.
-    </p>
-</description>
+<code-example [exampleFiles]="dropdownPopover"></code-example>\
```
Comment (katekozlowska) [inline apps/docs/src/app/core/component-docs/popover/popover-docs.component.html:198]: '\'  - remove leftover :)

**Review by katekozlowska — CHANGES_REQUESTED** [review]

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+        this._initialised = true;
+        this._refreshTriggerListeners();
+        if (this.isOpen) {
+            this.open();
+        }
+    }
+
+    /** @hidden */
+    ngOnChanges(changes: SimpleChanges): void {
+        if (!this._initialised) {
```
Comment (salarenko) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:141]: I think moving this to setters will save us from this additional "_initialised" flag.

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+
+    /** @hidden */
+    private _overlayRef: OverlayRef;
+
+    /** An RxJS Subject that will kill the data stream upon component’s destruction (for unsubscribing)  */
+    private readonly _onDestroy$: Subject<void> = new Subject<void>();
+
+    constructor(
+        private _renderer: Renderer2,
+        private _changeDetectorReference: ChangeDetectorRef,
```
Comment (salarenko) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:111]: I think we can shorten this to `_changeDetectorRef`

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+    private _listenOnClose(): void {
+        this._overlayRef.detachments()
+            .pipe(takeUntil(this._refresh$))
+            .subscribe(() => this.close())
+        ;
+    }
+
+    /** Listener for click events */
+    private _listenOnOutClicks(): void {
+        this._overlayRef.backdropClick().pipe(
```
Comment (salarenko) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:256]:
```
const closeEvents$ = merge( this._overlayRef.backdropClick(), this._overlayRef._outsidePointerEvents)

closeEvents$.pipe(
    filter(event => this._shouldClose(event)),
    takeUntil(this._refresh$)
).subscribe(() => this.close());
```

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    /** remove listeners from trigger element events */
+    private _removeTriggerListeners(): void {
+        this._eventRef.forEach(event => event());
+        this._eventRef = [];
+    }
+
+    /** @hidden */
+    private _triggerContainsTarget(event: Event): boolean {
+        const triggerElement = this._triggerElement.nativeElement;
+        return triggerElement.contains(event.composedPath()[0]);
```
Comment (salarenko) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:288]: IE11 crying because of not knowing what `composedPath` is

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+        }
+    }
+
+    /**
+     * Method called to change position of popover,
+     * recommended to be used only when popover is opened, otherwise change position or cdkPlacement
+     */
+    applyNewPosition(positions: ConnectedPosition[]): void {
+        const refPosition = this._getPositionStrategy(positions);
+        this._listenForPositionChange(refPosition.positionChanges);
```
Comment (salarenko) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:217]: Don't you have multiple listeners on `positionChange` after updating position?

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        this._overlayRef._outsidePointerEvents.pipe(
+            filter(event => this._shouldClose(event)),
+            takeUntil(this._refresh$)
+        ).subscribe(() => this.close());
+    }
+
+    /** Refresh listeners on trigger element events */
+    private _refreshTriggerListeners(): void {
+        this._removeTriggerListeners();
+        if (this.triggers && this.triggers.length > 0) {
```
Comment (salarenko) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:270]: Can be simplified to `if (this.triggers?.length > 0)` or even `if (this.triggers?.length)`

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+    /** Method called to refresh position of opened popover */
+    refreshPosition(): void {
+        if (this._overlayRef) {
+            this._overlayRef.updatePosition();
+        }
+    }
+
+    /** Handler for alt + arrow down keydown */
+    triggerKeyDownHandler(event: KeyboardEvent): void {
+        if (KeyUtil.isKey(event, 'ArrowDown') && event.altKey && !this.isOpen) {
```
Comment (salarenko) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:230]: Needs an update to latest `KeyUtil` changes

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+            this.open();
+        }
+    }
+
+    /** Handler escape keydown */
+    bodyKeydownHandler(event: KeyboardEvent): void {
+        if (KeyUtil.isKey(event, 'Escape') && this.isOpen && this.closeOnEscapeKey) {
```
Comment (salarenko) [inline libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:237]: Needs an update to latest `KeyUtil` changes

File: libs/core/src/lib/popover/base/base-popover.class.ts [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    /** Whether the popover is open. Can be used through two-way binding. */
+    @Input()
+    isOpen = false;
+
+    /** Whether the popover should close when a click is made outside its boundaries. */
+    @Input()
+    closeOnOutsideClick = true;
+
+    /** Event emitted when the state of the isOpen property changes. */
+    @Output()
```
Comment (salarenko) [inline libs/core/src/lib/popover/base/base-popover.class.ts:48]: Outputs after Inputs

**Review by salarenko — CHANGES_REQUESTED** [review]
Its a little bit confusing for me. How should we use `cdkPositions` and `placement` inputs together? The seem to define the same thing.

File: apps/docs/src/app/core/component-docs/popover/popover-docs.module.ts [hunk 14]
```diff
@@ -44,7 +55,11 @@ const routes: Routes = [
         DialogModule,
         ListModule,
         AvatarModule,
-        BarModule
+        BarModule,
+        SelectModule,
```
[the remark this answers is not in the document]
Comment (JKMarkowski) [reply apps/docs/src/app/core/component-docs/popover/popover-docs.module.ts:59]: Not anymore

File: libs/core/src/lib/combobox/combobox.component.ts [hunk 15]
```diff
@@ -480,6 +472,8 @@ export class ComboboxComponent implements ComboboxInterface, ControlValueAccesso
                 term: this.searchInputElement.nativeElement.value,
                 forceClose: false
             });
+
+            this.searchInputElement.nativeElement.focus();
```
[the remark this answers is not in the document]
Comment (JKMarkowski) [reply libs/core/src/lib/combobox/combobox.component.ts:489]: this shouldn't be there, moved to different place

File: libs/core/src/lib/datetime-picker/datetime-picker.component.ts [hunk 16]
```diff
@@ -560,22 +559,6 @@ export class DatetimePickerComponent implements OnInit, OnDestroy, ControlValueA
         }
     }
 
-    /** @hidden */
-    private _activateTimeComponent(): void {
```
[the remark this answers is not in the document]
Comment (JKMarkowski) [reply libs/core/src/lib/datetime-picker/datetime-picker.component.ts:564]: Yes :tada:

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts (hunk 5 above)
Comment (JKMarkowski) [reply libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:141]: Nope, setters will trigger open/close method, before view/content have been initialised.

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts (hunk 7 above)
Comment (JKMarkowski) [reply libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:256]: Thanks for catching it

File: libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts (hunk 9 above)
Comment (JKMarkowski) [reply libs/core/src/lib/popover/cdk-overlay/cdk-popover.component.ts:226]: I added some refresh, when this method is called

File: libs/core/src/lib/popover/popover-position/popover-position.ts [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    'top-end': { originX: 'end', originY: 'top', overlayX: 'end', overlayY: 'bottom' },
+    'bottom-start': { originX: 'start', originY: 'bottom', overlayX: 'start', overlayY: 'top' },
+    'bottom': { originX: 'center', originY: 'bottom', overlayX: 'center', overlayY: 'top' },
+    'bottom-end': { originX: 'end', originY: 'bottom', overlayX: 'end', overlayY: 'top' },
+    'left-start': { originX: 'start', originY: 'top', overlayX: 'end', overlayY: 'top' },
+    'left': { originX: 'start', originY: 'center', overlayX: 'end', overlayY: 'center' },
+    'left-end': { originX: 'start', originY: 'bottom', overlayX: 'end', overlayY: 'bottom' },
+    'right-start': { originX: 'end', originY: 'top', overlayX: 'start', overlayY: 'top' },
+    'right': { originX: 'end', originY: 'center', overlayX: 'start', overlayY: 'center' },
+    'right-end': { originX: 'end', originY: 'bottom', overlayX: 'start', overlayY: 'bottom' },
```
Comment (katekozlowska) [inline libs/core/src/lib/popover/popover-position/popover-position.ts:16]: remove coma

File: apps/docs/src/app/core/component-docs/popover/popover-docs.component.ts [hunk 18]
```diff
@@ -48,6 +57,64 @@ export class PopoverDocsComponent {
...
+            component: 'PopoverContainerExampleComponent',
+        }
+    ];
+
+    popoverTrigger: ExampleFile[] = [
+        {
+            language: 'html',
+            code: popoverTriggerSrc,
+            fileName: 'popover-trigger-example',
+            component: 'PopoverTriggerExampleComponent',
```
Comment (katekozlowska) [inline apps/docs/src/app/core/component-docs/popover/popover-docs.component.ts:74]: remove additional comas in this file

**Review by katekozlowska — DISMISSED** [review]

File: apps/docs/src/app/core/component-docs/popover/popover-docs.component.ts (hunk 18 above)
Comment (InnaAtanasova) [reply apps/docs/src/app/core/component-docs/popover/popover-docs.component.ts:74]: I think prettier is adding these...

**Comment by droshev** [comment]
@JKMarkowski when we deprecate something it is not yet a breaking change :)

File: libs/core/src/lib/popover/popover-position/popover-position.ts [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+        return _position;
+    }
+
+    static getMarginStyle(position: ArrowPosition): string {
+        if (!position) {
+            return;
+        }
+        if (position === 'start') {
+            return 'margin-left: 0.5rem;';
```
Comment (salarenko) [inline libs/core/src/lib/popover/popover-position/popover-position.ts:95]: Does not respects `ARROW_SIZE` for `start` and `end` position.

File: libs/core/src/lib/popover/popover.component.scss [hunk 20]
```diff
@@ -1,23 +1,125 @@
...
+$fd-popover-arrow-top-front: calc(-0.5rem + 1px) !default;
+
+$fd-popover-arrow-width: 1rem !default;
+$fd-popover-arrow-height: 0.5rem !default;
+$fd-popover-arrow-width-half: $fd-popover-arrow-width/2 !default;
+$fd-popover-arrow-offset: 0.5rem !default;
+$fd-popover-arrow-color: var(--sapContent_ForegroundBorderColor, #89919a) !default;
+$fd-popover-background-color: var(--sapGroup_ContentBackground, #fff) !default;
+
+.fd-popover__arrow {
```
Comment (salarenko) [inline libs/core/src/lib/popover/popover.component.scss:14]: What are these styles? Something custom?

**Review by salarenko — COMMENTED** [review]
- I thing its a good idea to have some directive for the trigger, like:
```
<button fd-button [popoverTrigger]="popover">Popover Trigger</button>

<fd-popover #popover>
    <div style="padding: 12px;">
        This popover is connected to trigger element
    </div>
</fd-popover>
```

and the directive takes care of providing `ElementRef`.

LGTM, last questions/suggestions.

File: libs/core/src/lib/popover/popover-position/popover-position.ts (hunk 19 above)
Comment (JKMarkowski) [reply libs/core/src/lib/popover/popover-position/popover-position.ts]: It has been improved. Thanks for catching it

File: libs/core/src/lib/popover/popover.component.scss (hunk 20 above)
Comment (JKMarkowski) [reply libs/core/src/lib/popover/popover.component.scss:14]: This is styling from `fundamental-styles`, but with changed selectors. Current style's css requires to keep `x-placement` attribute along with `fd-popper` class. Also now arrow sticks to middle of trigger element, instead body. It should be added in styles in feature.

**Comment by JKMarkowski** [comment]
> * I thing its a good idea to have some directive for the trigger, like:
> 
> ```
> <button fd-button [popoverTrigger]="popover">Popover Trigger</button>
> 
> <fd-popover #popover>
>     <div style="padding: 12px;">
>         This popover is connected to trigger element
>     </div>
> </fd-popover>
> ```
> 
> and the directive takes care of providing `ElementRef`.
> 
> LGTM, last questions/suggestions.

I added code almost same as on `menu-trigger.directive`. So now there is directive for trigger element

**Review by InnaAtanasova — APPROVED** [review]
